<div style="border:solid green 2px; padding: 20px">
<b>Hello, Valeriya!</b>
    
Grigoriy is here again, this is our final meeting :)  

You can find my comments in <font color='green'>green</font> or <font color='red'>red</font> boxes like this:

<div class="alert alert-success">
<b>Reviewer's comment:</b> if everything is done succesfully
</div>

<div class="alert alert-danger">
<b>Reviewer's comment:</b> if the block requires some corrections. Work can't be accepted with the red comments.
</div>
    
Let's work on the project in dialogue: if you change something in the project or respond to my comments, write about it. It will be easier for me to track the changes if you highlight your comments:
    
<div class="alert alert-info"> <b>Student's comment:</b> For example like this.</div>    
    
      
<div>

### <font color='orange'>General feedback</font>
Thank you for sending your project! Brilliant job, your research has been accepted!

---

# Final Project: SQL

### Name: Valeriya Vazhnova

### Email: valeria.vazh@gmail.com

## Goals of the Study

Following the coronavirus outbreak, people's lifestyle changed all around the world. Bound to spend their time indoors, people started reading more, hence the spike of popularity of various reading apps and book subscription services.

This study conducts a primary marketing research for a book-related startup. Luckily, we've managed to obtain databases of one of our competitors. Analysing this data will help us discover:  

- How active the book publishing business is;
- Who are the main players in this field: most successful publishers and most popular writers;
- How involved app users are;
- Which books have the highest rating and are often discussed (reviewed).


All this information will uncover the general trends in book publishing business, provide detailed recommendations for developing the new service and tailoring a new app for the market.

## <a id='cont'>Table of Contents</a>

[Connecting to the Database](#connecting)

[Studying the Datasets](#studying)

- [Books](#books)

- [Authors](#authors)

- [Publishers](#publishers)

- [Ratings](#ratings)

- [Reviews](#reviews)

[Analysing the Data](#analysing)

- [Overall Number of Books](#numberovbooks)

- [Number of Reviews and Average Rating](#reviewrating)

- [The Most Active Publisher](#activepublisher)

- [The Highest Rated Author](#highestratedauthor)

- [The Average Number of Text Reviews](#avgreviewsnumber)

[General Conclusions](#genconcl)

<div class="alert alert-success">
<b>Reviewer's comment:</b> 👍
</div>

## <a id='connecting'>Connecting to the Database</a>

In [1]:
# if needed

# pip install psycopg2-binary

In [2]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

# set up the connection through sqlalchemy

db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

[Back to Contents](#cont)

## <a id='studying'>Studying the Datasets</a>

We've been given five datasets with the information on the published books, their authors, publishers, ratings given by app users and text reviews written by app users.

### <a id='books'>Books</a>

In [3]:
query = '''
SELECT *
FROM books
LIMIT 10;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT COUNT (*) AS length
FROM books;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT COUNT (DISTINCT book_id) AS n_books
FROM books;'''

display(pd.io.sql.read_sql(query, con = engine))

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

length
0    1000

n_books
0     1000

**DESCRIPTION OF THE DATASET**

**Primary key:** book_id

**Foreign keys:** author_id, publisher_id

**Additional information:** title of the book, number of pages, publication date.

**Length:** 1000 unique entries

[Back to Contents](#cont)

### <a id='authors'>Authors</a>

In [4]:
query = '''
SELECT *
FROM authors
LIMIT 10;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT COUNT (*) AS length
FROM authors;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT COUNT (DISTINCT author_id) AS n_authors
FROM authors;'''

display(pd.io.sql.read_sql(query, con = engine))

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

length
0     636

n_authors
0        636

**DESCRIPTION OF THE DATASET**

**Primary key**: author_id

**Length:** 636 unique entries

[Back to Contents](#cont)



### <a id='publishers'>Publishers</a>

In [5]:
query = '''
SELECT *
FROM publishers
LIMIT 10;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT COUNT (*) AS length
FROM publishers;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT COUNT (DISTINCT publisher_id) AS n_publishers
FROM publishers;'''

display(pd.io.sql.read_sql(query, con = engine))

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

length
0     340

n_publishers
0           340

**DESCRIPTION OF THE DATASET**

**Primary key**: publisher_id

**Length:** 340 unique entries

**Warning**: We've noticed that imprints of the same publisher have different ids (e.g. Aladdin and Aladdin Paperbacks). Also there are clearly some dupliucates (Ace Book and Ace Books).

[Back to Contents](#cont)

### <a id='ratings'>Ratings</a>

In [6]:
query = '''
SELECT *
FROM ratings
LIMIT 10;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT COUNT (*) AS length
FROM ratings;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT
    COUNT (DISTINCT rating_id) AS n_ratings,
    COUNT (DISTINCT book_id) AS n_books
FROM ratings;'''

display(pd.io.sql.read_sql(query, con = engine))

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

length
0    6456

n_ratings  n_books
0       6456     1000

**DESCRIPTION OF THE DATASET**

**Primary key**: rating_id

**Foreign key:** book_id

**Length:** 6456 unique entries on 1000 books

[Back to Contents](#cont)

### <a id='reviews'>Reviews</a>

In [7]:
query = '''
SELECT *
FROM reviews
LIMIT 10;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT COUNT (*) AS length
FROM reviews;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT
    COUNT (DISTINCT review_id) AS n_reviews,
    COUNT (DISTINCT book_id) AS n_books
FROM reviews;'''

display(pd.io.sql.read_sql(query, con = engine))

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

length
0    2793

n_reviews  n_books
0       2793      994

**DESCRIPTION OF THE DATASET**

**Primary key:** review_id

**Foreign key:** book_id

**Length:** 2793 unique entries on 994 books

**Warning:** 

1) The total nubmer of text reviews is smaller than the number of ratings.

2) Not all the books have received text reviews.

3) In this dataset, reviews look gybberish, could it be auto-generated text?

[Back to Contents](#cont)

<div class="alert alert-success">
<b>Reviewer's comment:</b> Nicely done! Let's move on to the query part.
</div>

## <a id='analysing'>Analysing the Data</a>

### <a id='numberovbooks'>Overall Number of Books</a>

Find the number of books released after January 1, 2000.

In [8]:
query = '''
SELECT COUNT(DISTINCT book_id) AS n_books
FROM books
WHERE publication_date > '2000-01-01'
'''

pd.io.sql.read_sql(query, con = engine)

n_books
0      819

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

819 books from our dataset were published after 2020-01-01.

<div class="alert alert-info">
<b>Student's note:</b> As someone with a background in book publishing business, I know that this is astronomically far from the actual numbers. So here I can make two conclusions: either the competing app is rather small and unpopular, or the study data we're dealing with is far from real life.
</div>

[Back to Contents](#cont)

<div class="alert alert-success">
<b>Reviewer's comment: </b> Yes,  this is kind of toy dataset :)
</div>

### <a id='reviewrating'>Number of Reviews and Average Rating</a>

Find the number of user reviews and the average rating for each book.

In [9]:
# This query shows the first 15 rows. If you need the full table, please comment out (--) its last line (LIMIT 15;)

query = '''
SELECT
    DISTINCT reviews.book_id AS book_id,
    books.title,
    authors.author,
    COUNT(reviews.review_id) AS n_reviews,
    subquery.avg_rating,
    subquery.n_ratings
FROM reviews
LEFT JOIN
    (
    SELECT 
        DISTINCT book_id,
        AVG(rating) AS avg_rating,
        COUNT(DISTINCT rating_id) AS n_ratings
    FROM ratings 
    GROUP BY book_id
    ) AS subquery ON subquery.book_id = reviews.book_id
LEFT JOIN books ON books.book_id = reviews.book_id
LEFT JOIN authors ON authors.author_id = books.author_id
GROUP BY
    reviews.book_id,
    books.title,
    authors.author,
    subquery.avg_rating,
    subquery.n_ratings
ORDER BY n_reviews DESC
LIMIT 15;
'''

pd.io.sql.read_sql(query, con = engine)

book_id                                              title  \
0       948                            Twilight (Twilight  #1)   
1       733                          The Giver (The Giver  #1)   
2       497                          Outlander (Outlander  #1)   
3       627                                      The Alchemist   
4       779  The Lightning Thief (Percy Jackson and the Oly...   
5       673                             The Catcher in the Rye   
6       750                The Hobbit  or There and Back Again   
7       656                                     The Book Thief   
8       207                                    Eat  Pray  Love   
9       299  Harry Potter and the Chamber of Secrets (Harry...   
10      696             The Da Vinci Code (Robert Langdon  #2)   
11      854                                           The Road   
12      963                                Water for Elephants   
13      302  Harry Potter and the Prisoner of Azkaban (Harr...   
14      734                                   The Glass Castle   

                                      author  n_reviews  avg_rating  n_ratings  
0                            Stephenie Meyer          7    3.662500        160  
1                                 Lois Lowry          6    3.750000         56  
2                             Diana Gabaldon          6    4.125000         24  
3   Paulo Coelho/Alan R. Clarke/Özdemir İnce          6    3.789474         57  
4                               Rick Riordan          6    4.080645         62  
5                              J.D. Salinger          6    3.825581         86  
6                             J.R.R. Tolkien          6    4.125000         88  
7          Markus Zusak/Cao Xuân Việt Khương          6    4.264151         53  
8                          Elizabeth Gilbert          6    3.395833         48  
9                 J.K. Rowling/Mary GrandPré          6    4.287500         80  
10                                 Dan Brown          6    3.830508         59  
11                           Cormac McCarthy          6    3.772727         22  
12                                Sara Gruen          6    3.977273         44  
13                J.K. Rowling/Mary GrandPré          6    4.414634         82  
14                           Jeannette Walls          6    4.206897         29

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

These are the top-15 books based on the number of text reviews they got and their average rating. Among them we see both recent bestsellers (Twilight, Harry Potter series, The Alchemist, Eat Pray Love etc.) and older books which by now can be considered classics of their genre (The Catcher in the Rye, The Hobbit or There and Back Again).

In [10]:
# this function contains most of the query, and lets you display various parts of the dataset (beginning, end,
# ordered by different column etc.)

def reviews_rating_query(end_lines):
    query_text = '''
    SELECT
        DISTINCT reviews.book_id AS book_id,
        books.title,
        authors.author,
        COUNT(reviews.review_id) AS n_reviews,
        subquery.avg_rating,
        subquery.n_ratings
    FROM reviews
    LEFT JOIN
        (
        SELECT 
            DISTINCT book_id,
            AVG(rating) AS avg_rating,
            COUNT(DISTINCT rating_id) AS n_ratings
        FROM ratings 
        GROUP BY book_id
        ) AS subquery ON subquery.book_id = reviews.book_id
    LEFT JOIN books ON books.book_id = reviews.book_id
    LEFT JOIN authors ON authors.author_id = books.author_id
    GROUP BY
        reviews.book_id,
        books.title,
        authors.author,
        subquery.avg_rating,
        subquery.n_ratings
    '''
    full_query = query_text+end_lines
    return display(pd.io.sql.read_sql(full_query, con = engine))

# by the smallest number of reviews
print('Books with the smallest number of reveiws')
end_lines = '''
ORDER BY n_reviews
LIMIT 10;
'''
reviews_rating_query(end_lines)
print('')

# by the highest average rating
print('Books with the highest average rating')
end_lines = '''
ORDER BY avg_rating DESC
LIMIT 10;
'''
reviews_rating_query(end_lines)
print('')

# by the lowest average rating
print('Books with the lowest average rating')
end_lines = '''
ORDER BY avg_rating
LIMIT 10;
'''
reviews_rating_query(end_lines)
print('')


Books with the smallest number of reveiws


book_id                                              title  \
0      314                                           Hideaway   
1      129  Built to Last: Successful Habits of Visionary ...   
2      431                Merrick (The Vampire Chronicles #7)   
3      669       The Calvin and Hobbes Tenth Anniversary Book   
4      980                              Wicked: The Grimmerie   
5      446                                Moo  Baa  La La La!   
6      943               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1   
7      639                                      The Bad Place   
8      318                                  History of Beauty   
9      617                              Ten Apples Up On Top!   

                                              author  n_reviews  avg_rating  \
0                                        Dean Koontz          1         4.0   
1                   James C. Collins/Jerry I. Porras          1         4.5   
2                                          Anne Rice          1         4.0   
3                                     Bill Watterson          1         3.5   
4  David Cote/Stephen Schwartz/Joan Marcus/Winnie...          1         3.5   
5                                     Sandra Boynton          1         3.0   
6                               CLAMP/Anthony Gerard          1         2.5   
7                                     Dean R. Koontz          1         3.5   
8  Umberto Eco/Girolamo De Michele/Alastair McEwe...          1         2.5   
9                    Theo LeSieg/Dr. Seuss/Roy McKie          1         4.0   

   n_ratings  
0          2  
1          2  
2          2  
3          2  
4          2  
5          2  
6          2  
7          2  
8          2  
9          2


Books with the highest average rating


book_id                                              title  \
0      275                                          Geek Love   
1      610                           Tai-Pan (Asian Saga  #2)   
2      444       Moneyball: The Art of Winning an Unfair Game   
3      169  Crucial Conversations: Tools for Talking When ...   
4      224                                      Evening Class   
5      699  The Demon-Haunted World: Science as a Candle i...   
6      513               Piercing the Darkness (Darkness  #2)   
7      516                       Plum Island (John Corey  #1)   
8      441                  Misty of Chincoteague (Misty  #1)   
9       57                    Act of Treason (Mitch Rapp  #9)   

                                              author  n_reviews  avg_rating  \
0                                     Katherine Dunn          2         5.0   
1                                      James Clavell          2         5.0   
2                                    Michael   Lewis          3         5.0   
3  Kerry Patterson/Joseph Grenny/Ron McMillan/Al ...          2         5.0   
4                                       Maeve Binchy          2         5.0   
5                              Carl Sagan/Ann Druyan          2         5.0   
6                                   Frank E. Peretti          2         5.0   
7                                     Nelson DeMille          2         5.0   
8                     Marguerite Henry/Wesley Dennis          2         5.0   
9                                        Vince Flynn          2         5.0   

   n_ratings  
0          2  
1          2  
2          3  
3          2  
4          2  
5          2  
6          2  
7          2  
8          2  
9          2


Books with the lowest average rating


book_id                                              title  \
0      303                               Harvesting the Heart   
1      202                                      Drowning Ruth   
2      316                  His Excellency: George Washington   
3      371                                              Junky   
4      915  The World Is Flat: A Brief History of the Twen...   
5      794                                  The Mermaid Chair   
6      772                             The Kitchen God's Wife   
7      822                                     The Penelopiad   
8      983  Wild at Heart: Discovering the Secret of a Man...   
9      470          Neither Here nor There: Travels in Europe   

                                              author  n_reviews  avg_rating  \
0                                       Jodi Picoult          2    1.500000   
1                                  Christina Schwarz          3    2.000000   
2                                    Joseph J. Ellis          2    2.000000   
3  William S. Burroughs/Oliver  Harris/Allen Gins...          2    2.000000   
4                                 Thomas L. Friedman          3    2.250000   
5                                      Sue Monk Kidd          3    2.333333   
6                                            Amy Tan          3    2.333333   
7                  Margaret Atwood/Laurel Merlington          2    2.500000   
8                                      John Eldredge          2    2.500000   
9                                        Bill Bryson          2    2.500000   

   n_ratings  
0          2  
1          3  
2          2  
3          2  
4          4  
5          3  
6          3  
7          2  
8          2  
9          2

<a id='importanceofnumber'></a>

Interesting to notice that number of reviews and average rating don't seem to be strongly correlated. We see both books with high average rating but the small number of reviews (e.g. To Green Angel Tower by Tad Williams) as well as books with mediocre rating and high number of reviews (e.g. Twilight by Stephenie Meyer).

The key is, the less people rate the book, the more weight each individual rating has. **So when trying to assess the book by its average rating, we suggest you should also pay attention to the number of people who rated it.**

Perhaps, a good parameter to assess the book would be the ratio of reviews per ratings.

[Back to Contents](#cont)

### <a id='activepublisher'>The Most Active Publisher</a>

Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [11]:
# This query shows only the first row. If you need the full table, please comment out (--) its last line (LIMIT 1;)

query = '''
SELECT
    books.publisher_id,
    publishers.publisher,
    COUNT(books.book_id) AS books_count
FROM books
LEFT JOIN publishers on publishers.publisher_id = books.publisher_id
WHERE num_pages > 50
GROUP BY
    books.publisher_id,
    publishers.publisher
ORDER BY books_count DESC
LIMIT 1;
'''

pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  books_count
0           212  Penguin Books           42

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

So, the most active publisher is Penguin Books with 42 books in our dataset. This doesn't come as a surprise, since it's a big name in book industry worlwide. We simply want to remind you that this calculation doesn't include other imprints of Penguin Random House UK (such as Penguin Classics, Vintage etc.).

[Back to Contents](#cont)

### <a id='highestratedauthor'>The Highest Rated Author</a>

Identify the author with the highest average book rating: look only at books with at least 50 ratings.

In [12]:
# This function contains most of the query, and lets you display various parts of the dataset (beginning, end,
# ordered by different column etc.)

def popular_author_query(end_lines):
    query_text = '''
    SELECT
        books.author_id,
        authors.author,
        COUNT(ratings.rating_id) AS n_ratings,
        AVG(ratings.rating) AS avg_rating
    FROM ratings
    LEFT JOIN books ON books.book_id = ratings.book_id
    LEFT JOIN authors ON authors.author_id = books.author_id
    GROUP BY
        books.author_id,
        authors.author
    '''
    full_query = query_text+end_lines
    return display(pd.io.sql.read_sql(full_query, con = engine))
    
end_lines='''
HAVING COUNT(ratings.rating_id) >= 50
ORDER BY avg_rating DESC
LIMIT 5;
'''
print('Ordered by average rating')
popular_author_query(end_lines)
print('')
    
end_lines='''
HAVING COUNT(ratings.rating_id) >= 50
ORDER BY
    n_ratings DESC,
    avg_rating DESC
LIMIT 5;
'''
print('Ordered by number of ratings and average rating')
popular_author_query(end_lines)
print('')

end_lines='''
HAVING COUNT(ratings.rating_id) >= 100
ORDER BY avg_rating DESC
LIMIT 5;
'''
print('Authors with 100 ratings or higher (ordered by average rating)')
popular_author_query(end_lines)
print('')

Ordered by average rating


author_id                             author  n_ratings  avg_rating
0        130                     Diana Gabaldon         50    4.300000
1        236         J.K. Rowling/Mary GrandPré        312    4.288462
2          3                    Agatha Christie         53    4.283019
3        402  Markus Zusak/Cao Xuân Việt Khương         53    4.264151
4        240                     J.R.R. Tolkien        166    4.240964


Ordered by number of ratings and average rating


author_id                      author  n_ratings  avg_rating
0        236  J.K. Rowling/Mary GrandPré        312    4.288462
1        240              J.R.R. Tolkien        166    4.240964
2        554             Stephenie Meyer        160    3.662500
3        106                   Dan Brown        143    3.741259
4        445             Nicholas Sparks        111    3.882883


Authors with 100 ratings or higher (ordered by average rating)


author_id                      author  n_ratings  avg_rating
0        236  J.K. Rowling/Mary GrandPré        312    4.288462
1        240              J.R.R. Tolkien        166    4.240964
2        542                Stephen King        106    4.009434
3        445             Nicholas Sparks        111    3.882883
4        106                   Dan Brown        143    3.741259

This is a direct consequence of [the fact we mentioned earlier](#importanceofnumber): the number of people who rated a book is extremely important.

If we calculate just the average book rating for every author, the highest rated writer is suddenly Diana Gabaldon, author of romance fantasy novels. However, this score doesn't reflect the real sutuation: it's based on 50 reviews only and means that Ms. Gabaldon has a small, but devoted fan club who rate all her books pretty high.

However, if we additionally order the table by the number of ratings (or simply set a higher entry level), we see the real winner of this book race, **J.K. Rowling**.

We'll achieve the same result also if we calculate average rating not per all books of the author, but per each single book.

In [13]:
# This query shows only the first row. If you need the full table, please comment out (--) its last line (LIMIT 1;)

query = '''
SELECT
    books.author_id,
    books.book_id,
    books.title,
    authors.author,
    COUNT(ratings.rating_id) AS n_ratings,
    AVG(ratings.rating) AS avg_rating
FROM ratings
LEFT JOIN books ON books.book_id = ratings.book_id
LEFT JOIN authors ON authors.author_id = books.author_id
GROUP BY
    books.author_id,
    books.book_id,
    books.title,
    authors.author
HAVING COUNT(ratings.rating_id) >= 50
ORDER BY avg_rating DESC
LIMIT 1;
'''

pd.io.sql.read_sql(query, con = engine)

author_id  book_id                                              title  \
0        236      302  Harry Potter and the Prisoner of Azkaban (Harr...   

                       author  n_ratings  avg_rating  
0  J.K. Rowling/Mary GrandPré         82    4.414634

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

[Back to Contents](#cont)

### <a id='avgreviewsnumber'>The Average Number of Text Reviews</a>

Find the average number of text reviews among users who rated more than 50 books.

In [14]:
# average number

query = '''
SELECT AVG(n_reviews)
FROM
    (
    SELECT COUNT(DISTINCT review_id) AS n_reviews
    FROM reviews
    WHERE username IN
        (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(DISTINCT book_id) > 50
        )
    GROUP BY username) AS subquery;
'''

pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

In [15]:
# in details: how many reviews wrote each of these users

query = '''
SELECT
    username,
    COUNT(DISTINCT review_id) AS n_reviews
FROM reviews
WHERE username IN
        (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(DISTINCT book_id) > 50
        )
GROUP BY username
ORDER BY n_reviews DESC;
'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT
    COUNT (DISTINCT username) AS n_reviewers
FROM reviews;'''

display(pd.io.sql.read_sql(query, con = engine))

query = '''
SELECT
    COUNT (DISTINCT username) AS n_users_who_rate_books
FROM ratings;'''

display(pd.io.sql.read_sql(query, con = engine))

username  n_reviews
0     sfitzgerald         28
1      martinadam         27
2       richard89         26
3  jennifermiller         25
4          paul88         22
5          xdavis         18

n_reviewers
0          160

n_users_who_rate_books
0                     160

These six people are the most active users who rated more than 50 books and wrote quite a lot of reviews: the average number is 24 for each of them. However, the total number of users who write reviews and rate books is 160.

In [16]:
print('Only {:.2%} of the app users contribute actively.'.format(6 / 160))

Only 3.75% of the app users contribute actively.


So, unfortunately, we can declare that most of the app users are not that active in generating content.

[Back to Contents](#cont)

## <a id='genconcl'>General Conclusions</a>

We've analysed the given datasets and came to certain conclusions which would be really improtant for those launching a book related app.

1) When selecting books for the app, you shouldn't rely solemnly on their rating. The rating itself is unrepresentative. You should always understand how many single grades it's based on.

2) You shouldn't expect app users to be very active. Only a small fraction of them generates content actively (leaves reviews, rates books). In this dataset, the share of active users was only 3.75%. If you expect user-generated content in your app, you should encourage it by offering some kind of promotions or discounts.

3) All the top-15 list was occupied by fiction books, both classics and recent bestsellers. We've noticed that our dataset contains several nonfiction books, but they have either low average rating or very little reviews. However, simply saying that fiction is more popular than nonfiction would be too far-fetched. It's possible that our dataset has a bigger share of fiction and lacks the biggest nonfiction names (Steven Hawking, Richard Dawkins, Dale Carnegie, Naomi Wolf etc.). With different selection of names and genres in the given dataset we could've come to a different conclusion.

4) And finally, some names you shouldn't miss out on: **Penguin Books** as one of the most popular and active book publishers, and **J.K. Rowling** as the highest rated author.

[Back to Contents](#cont)

<div class="alert alert-success">
<b>Reviewer's comment:</b> Thank you for so comprehensive research! You've definately gone the extra mile :)
</div>
